### PR TITLE
fixed problem with implicit declaration in quantum/rgblight.c

### LINKED ISSUE #8381: [Bug] implicit declaration of function
I'm trying to implement layer-base underglow.

When i import `#define RGBLIGHT_LAYERS` i get a bunch of compile errors.
## Describe the Bug
I am following https://docs.qmk.fm/#/feature_rgblight
I added the following to `config.h`:
```
#define RGBLIGHT_LAYERS
```
and naturally the following to `rules.mk`:
```
RGBLIGHT_ENABLE = yes
```
but upon make'ing, iget the following errors - which seems to be errors in the framework itself (and not due to my implementation, as it doesn't bug out without the `#define RGBLIGHT_LAYERS`):
```
Compiling: quantum/rgblight.c                                                                      quantum/rgblight.c: In function 'rgblight_layers_write':
quantum/rgblight.c:645:49: error: implicit declaration of function 'pgm_read_ptr'; did you mean 'pgm_read_word'? [-Werror=implicit-function-declaration]
  645 |         const rgblight_segment_t *segment_ptr = pgm_read_ptr(layer_ptr);
      |                                                 ^~~~~~~~~~~~
      |                                                 pgm_read_word
quantum/rgblight.c:645:49: error: initialization of 'const rgblight_segment_t *' {aka 'const struct <anonymous> *'} from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
quantum/rgblight.c:652:13: error: implicit declaration of function 'memcpy_P'; did you mean 'memcpy'? [-Werror=implicit-function-declaration]
  652 |             memcpy_P(&segment, segment_ptr, sizeof(rgblight_segment_t));
      |             ^~~~~~~~
      |             memcpy
cc1: all warnings being treated as errors
 [ERRORS]
```

I tried different things and among others `make mxss:default`, which gave me header errors as well:
```
Compiling: keyboards/mxss/rgblight.c                                                               In file included from keyboards/mxss/rgblight.c:30:
keyboards/mxss/rgblight.h:19: error: "pgm_read_word" redefined [-Werror]
   19 | #define pgm_read_word(addr) (*(const unsigned short *)(addr))
      |
In file included from tmk_core/common/progmem.h:4,
                 from keyboards/mxss/rgblight.c:28:
/usr/avr/include/avr/pgmspace.h:991: note: this is the location of the previous definition
  991 | #define pgm_read_word(address_short)    pgm_read_word_near(address_short)
      |
In file included from keyboards/mxss/rgblight.c:30:
keyboards/mxss/rgblight.h:20:10: fatal error: rgblight_reconfig.h: No such file or directory
   20 | #include "rgblight_reconfig.h"
      |          ^~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
compilation terminated.
 [ERRORS]
```
the reconfig file seems, though, to be the file that would define the missing `pgm_read_ptr` function. So these bugs seem related. 

I wanted to dig into the source and provide an actual bugfix but i'm not that great at C and don't know your codebase well, so i would be pretty useless. Also it might just be my config or something that is bad.

For reference, my keymap can be found here:

https://github.com/C0DK/dotfiles/tree/master/keymaps/cabang

## System Information

 - Keyboard:  
   - Revision (if applicable):  Preonic Rev3
 - Operating system: 4.19.107-1-MANJARO
 - AVR GCC version: `9.2.0`
 - ARM GCC version: `9.2.0`
 - QMK Firmware version: `0.8.26`

### PR DESCRIPTION
**Pull request by C0DK**
## Description

So i am well aware that my solution probably isn't up to standards - neither code nor modularity nor DRY principles - i don't normally work with C, but had a bug that made the layers stuff useless. I fixed this for my working version and would just, as a friendly gesture, provide you guys with the solution. I hope you can possibly guide me to a better way of fixing the bug (maybe by importing the correct header, or something else)

I also want to say that this introduces a warning, which i had to ignore, but it works fine. 

I hope you can help me help you, and hopefully i can learn something in the go.

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Fixes #8381

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by fauxpark** [comment]
These declarations would be better placed in tmk_core/common/progmem.h - on AVR all this header does is include <avr/pgmspace.h>, but for ARM, RAM and program memory are on the same bus, so these `_P` functions need to be aliased to the normal ones.

**Comment by C0DK** [comment]
@fauxpark So i'm not sure how to do exactly that. these lines are copied from other files, but those files were (presumably) not included in this - however i do not know how includes work in these files. I'm not that sharp in C (pun very much intended)

**Comment by fauxpark** [comment]
Since `progmem.h` is already included at the top of `rgblight.c`, all that needs to be done is add to its contents (or rather, replace the whole thing) as follows:
```c
#pragma once

#if defined(__AVR__)
#    include <avr/pgmspace.h>
#else
#    define PROGMEM
#    define memcpy_P(dest, src, n) memcpy(dest, src, n)
#    define pgm_read_byte(address_short) *((unsigned char*)(address_short))
#    define pgm_read_dword(address_short) *((uint32_t*)(address_short))
#    define pgm_read_word(address_short) *((uint16_t*)(address_short))
#    define pgm_read_ptr(address_short) *((void*)(address_short))
#    define strcmp_P(s1, s2) strcmp(s1, s2)
#    define strcpy_P(dest, src) strcpy(dest, src)
#    define strlen_P(src) strlen(src)
#endif
```
I've thrown in a couple extra `_P` aliases I found from grepping the repo - mostly in the Bluetooth code and will come in handy for the ARM port. May as well kill two birds with one stone 😄
I also changed the argument names to match what they are in (avr-)libc. Won't affect functionality.

**Comment by C0DK** [comment]
You are the man! i'll change it today if possible. Would that be all the changes needed? i'll be a major win if i get some code in production on such a big project (Mum i did it!)

**Comment by fauxpark** [comment]
I almost forgot, the `memcpy_P` alias should also be removed from `drivers/oled/oled_driver.c` otherwise the compiler will complain about the multiple definitions.

https://github.com/qmk/qmk_firmware/blob/b9d0b1f06478a1b3f596882b33cb9bd84e477bea/drivers/oled/oled_driver.c#L26-L28

**Comment by C0DK** [comment]
There we go. I moved the changes as proposed. Thanks :) @fauxpark

**Comment by fauxpark** [comment]
Almost, the changes to progmem.h aren't in there.

**Comment by C0DK** [comment]
> Almost, the changes to progmem.h aren't in there.

Damn you have to press "save". 

Here we go.

File: tmk_core/common/progmem.h [hunk 1]
```diff
@@ -4,7 +4,12 @@
 #    include <avr/pgmspace.h>
 #else
 #    define PROGMEM
+#    define memcpy_P(dest, src, n) memcpy(dest, src, n)
 #    define pgm_read_byte(p) *((unsigned char*)(p))
 #    define pgm_read_word(p) *((uint16_t*)(p))
 #    define pgm_read_dword(p) *((uint32_t*)(p))
```
Comment (fauxpark) [inline tmk_core/common/progmem.h]:
```suggestion
#    define pgm_read_byte(address_short) *((uint8_t*)(address_short))
#    define pgm_read_word(address_short) *((uint16_t*)(address_short))
#    define pgm_read_dword(address_short) *((uint32_t*)(address_short))
```
For consistency :)

**Comment by zvecr** [comment]
Thanks!

**Pull request merged**